core-bundle/src/Resources/contao/modules/ModulePassword.php

```diff
@@ -319,11 +320,11 @@ protected function sendPasswordLink($objMember)
 		$arrData['domain'] = Idna::decode(Environment::get('host'));
 		$arrData['link'] = Idna::decode(Environment::get('base')) . Environment::get('request') . ((strpos(Environment::get('request'), '?') !== false) ? '&' : '?') . 'token=' . $optInToken->getIdentifier();
 
+		$text = $container->get('contao.string.simple_token_parser')->parse($this->reg_password, $arrData);
+		$text = $container->get('contao.insert_tag.parser')->replaceInline($text);
```

Comment:
```suggestion
		$text = $container->get('contao.insert_tag.parser')->replaceInline($this->reg_password);
		$text = $container->get('contao.string.simple_token_parser')->parse($text, $arrData);
```

If we flip these calls, injecting insert tags via untrused input should no longer be possible.

Reply:
The problem is, that people would want to use `{{foobar::##whatever##}}` - the other way around makes no sense.

Reply:
But then something like `{{date_format::##lastLogin##::d.m.Y}}` will not be possible unfortunately (see also https://github.com/contao/contao/pull/1768).

Reply:
What about we prevent either the simple token parser to return anything that begins with `{{` or insert tags to return something that begins with `##`? Or both? It's highly unlikely that you ever want this on purpose, I think.

Reply:
As long as `{{foobar::##whatever##}}` would return the input-string somewhere, the replacement would still work.

e.g.:
1. Original: `{{hello::##input##}}`
2. Insert tags replaced: `Hello ##input##, how are you?`
3. Simple tokens replaced: `Hello Martin, how are you?`

Reply:
Yes but you want the parsed value to be passed on to the insert tag - see @fritzmg's example.

Reply:
How about changing https://github.com/contao/contao/blob/09b21399b1a4e82019676d2502763105d461ec22/core-bundle/src/Resources/contao/library/Contao/Input.php#L967 to:

```php
return str_replace(array('{', '}'), array('&#123;', '&#125;'), (string) $varValue); 
```

This way all `{` and `}` characters would be encoded and it should be impossible to inject an insert tag this way.

Reply:
And what if you did not use the `Input` class - or won't use it in the future maybe? Maybe we need a service that does both, handling insert tags together with simple tokens in a safe way? I mean, it kind of sounds like a fragile API design if you have to know which one of those 2 services to apply first so that it's safe 🤔 I think one would always want to parse the simple tokens first and then the insert tags - I cannot think of why you would want that otherwise.

Reply:
it is all very unrelated to this ticket though. Its something we should improve, but currently parsing as in the PR is imho what we always do in Contao 🤷

Reply:
No, we cannot merge this for security reasons so we have to find another solution first.

Reply:
I don't get it. You're saying it is insecure to use insert tags and simple tokens in an email text, but it is perfectly safe to use them all over websites (e.g. on a form confirmation page)???